core/src/main/java/hudson/model/AbstractBuild.java

```diff
@@ -723,21 +725,33 @@ protected final boolean performAllBuildSteps(BuildListener listener, Iterable<?
                     try {
                         if (!perform(bs,listener)) {
                             LOGGER.log(Level.FINE, "{0} : {1} failed", new Object[] {AbstractBuild.this, bs});
+                            setResult(Result.FAILURE);
                             r = false;
                         }
                     } catch (Exception e) {
                         reportError(bs, e, listener, phase);
+                        r = false;
                     } catch (LinkageError e) {
                         reportError(bs, e, listener, phase);
+                        r = false;
```

Comment:
`performAllBuildSteps` is ignored in most calls, but i think this should add consistency for failed execution